### PR TITLE
OG-193: Relay Server refactoring

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: src/relayclient/ContractInteractor.ts [hunk 1]
```diff
@@ -333,6 +330,49 @@ export default class ContractInteractor {
     const code = await this.web3.eth.getCode(address)
     return code !== '0x'
   }
+
+  async getStakeInfo (managerAddress: Address): Promise<{
```
Comment (shahafn) [inline src/relayclient/ContractInteractor.ts:334]: see OG-192 to fix it

File: src/relayserver/RegistrationManager.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    receipts = receipts.concat((await this.transactionManager.sendTransaction({
+      signer: this.managerAddress,
+      method: registerMethod,
+      destination: this.hubAddress
+    })).receipt)
+    log.debug(`Relay ${this.managerAddress} registered on hub ${this.hubAddress}. `)
+    return receipts
+  }
+
+  async _sendMangerEthBalanceToOwner (): Promise<TransactionReceipt[]> {
```
Comment (shahafn) [inline src/relayserver/RegistrationManager.ts:264]: typo in `Manger` (missing 'a')

File: src/relayserver/RelayServer.ts [hunk 3]
```diff
@@ -232,19 +102,19 @@ export class RelayServer extends EventEmitter {
...
     const workerIndex = 0
-    if (req.relayWorker.toLowerCase() !== this.getWorkerAddress(workerIndex).toLowerCase()) {
+    if (req.relayWorker.toLowerCase() !== this.workerAddress.toLowerCase()) {
       throw new Error(
         `Wrong worker address: ${req.relayWorker}\n`)
     }
 
     // if trusted paymaster, we trust it to handle fees
-    if (!this.trustedPaymasters.includes(req.paymaster.toLowerCase())) {
+    if (!this.config.trustedPaymasters.map(it => it.toLowerCase()).includes(req.paymaster.toLowerCase())) {
```
Comment (shahafn) [inline src/relayserver/RelayServer.ts:111]: Can you do the lowercasing in the constructor?

File: src/relayserver/RelayServer.ts [hunk 4]
```diff
@@ -417,72 +291,81 @@ export class RelayServer extends EventEmitter {
   }
 
   async _init (): Promise<void> {
+    // TODO: add rerun protection
```
Comment (shahafn) [inline src/relayserver/RelayServer.ts:294]: `if (this.initialized) {
      return
  }`

File: src/relayserver/TransactionManager.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+      const txToSign = new Transaction({
+        to: destination,
+        value: value,
+        gasLimit: gas,
+        gasPrice: _gasPrice,
+        data: Buffer.from(encodedCall.slice(2), 'hex'),
+        nonce
+      }, this.rawTxOptions)
+      log.trace('txToSign', txToSign)
+      // TODO omg! do not do this!
```
Comment (shahafn) [inline src/relayserver/TransactionManager.ts:82]: So fix...

File: src/relayserver/TransactionManager.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+      const keyManager = this.managerKeyManager.isSigner(signer) ? this.managerKeyManager : this.workersKeyManager
+      signedTx = keyManager.signTransaction(signer, txToSign)
+      storedTx = transactionToStoredTx(txToSign, signer, 1)
+      this.nonces[signer]++
+      await this.txStoreManager.putTx(storedTx, false)
+    } finally {
+      releaseMutex()
+    }
+    const receipt = await this.contractInteractor.sendSignedTransaction(signedTx)
+    log.debug('\ntxhash is', receipt.transactionHash)
```
Comment (shahafn) [inline src/relayserver/TransactionManager.ts:92]: This should be info imo, logging tx hash for each transaction sent by server is reasonable.

File: src/relayserver/TransactionManager.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    log.debug('txToSign', txToSign)
+    const keyManager = this.managerKeyManager.isSigner(tx.from) ? this.managerKeyManager : this.workersKeyManager
+    const signedTx = keyManager.signTransaction(tx.from, txToSign)
+    const storedTx = transactionToStoredTx(txToSign, tx.from, tx.attempts + 1)
+    await this.txStoreManager.putTx(storedTx, true)
+
+    log.debug('resending tx with nonce', txToSign.nonce, 'from', tx.from)
+    log.debug('account nonce', await this.contractInteractor.getTransactionCount(tx.from))
+    const receipt = await this.contractInteractor.sendSignedTransaction(signedTx)
+    log.debug('\ntxhash is', receipt.transactionHash)
```
Comment (shahafn) [inline src/relayserver/TransactionManager.ts:130]: Maybe that should also be info when resending tx with new txhash?

File: test/relayserver/RelayServer.test.ts [hunk 8]
```diff
@@ -857,100 +367,172 @@ contract('RelayServer', function (accounts) {
...
+      const refill = toBN(relayServer.config.workerTargetBalance).sub(workerBalanceBefore)
       assert.isTrue(managerHubBalanceBefore.lt(refill), 'manager hub balance should be insufficient to replenish worker')
       assert.isTrue(managerEthBalance.lt(refill), 'manager eth balance should be insufficient to replenish worker')
       let fundingNeededEmitted = false
       relayServer.on('fundingNeeded', () => { fundingNeededEmitted = true })
       await relayServer.replenishServer(workerIndex)
       assert.isTrue(fundingNeededEmitted, 'fundingNeeded not emitted')
     })
+
+    // TODO: unskip these tests. They are currently integration tests for server <-> registration manager
```
Comment (shahafn) [inline test/relayserver/RelayServer.test.ts:487]: Please unskip them in this PR, create a new test file for this integration if you'd like to...

File: test/relayserver/TransactionManager.test.ts [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+import { GsnRequestType } from '../../src/common/EIP712/TypedRequestData'
+import { GSNConfig } from '../../src/relayclient/GSNConfigurator'
+
+const TestPaymasterEverythingAccepted = artifacts.require('TestPaymasterEverythingAccepted')
+const TestRecipient = artifacts.require('TestRecipient')
+const StakeManager = artifacts.require('StakeManager')
+const Penalizer = artifacts.require('Penalizer')
+const Forwarder = artifacts.require('Forwarder')
+
+contract('TransactionManager', function (accounts) {
```
Comment (shahafn) [inline test/relayserver/TransactionManager.test.ts:31]: If we're at it, this should also cover `sendTransaction()` missing tests... should open jira issue for that at least

File: src/relayserver/RegistrationManager.ts (hunk 2 above)
Comment (forshtat) [reply src/relayserver/RegistrationManager.ts:264]: fixed

File: src/relayserver/RelayServer.ts (hunk 3 above)
Comment (forshtat) [reply src/relayserver/RelayServer.ts:111]: No, as I don't touch the 'config' in the constructor, only assign it. 
Having to lowercase addresses in constructor would be awkward.

File: src/relayserver/RelayServer.ts (hunk 4 above)
Comment (forshtat) [reply src/relayserver/RelayServer.ts:294]: fixed

File: src/relayserver/TransactionManager.ts (hunk 6 above)
Comment (forshtat) [reply src/relayserver/TransactionManager.ts:92]: Will fix when working on the multiple workers

Comment (forshtat) [reply src/relayserver/TransactionManager.ts:92]: fixed

File: src/relayserver/TransactionManager.ts (hunk 7 above)
Comment (forshtat) [reply src/relayserver/TransactionManager.ts:130]: fixed

**Review by shahafn — APPROVED** [review]

**Pull request merged**